### PR TITLE
fix: manually update apollo cache on activity completion

### PR DESCRIPTION
**Pull request by mohsinht**
- Sometimes the cache is not properly updated in case of form submission or checklist submission. The same completed activity is shown to the user again. We've been observing this in DataDog tests and Checkly. To solve this issue, I've added a hook that forces the Apollo Cache to update the session activities manually. ([see loom](https://www.loom.com/share/3a4d04235de949caa7f145ef468e873e?sid=fcd381c1-e7cb-4a99-a478-50c7c7334142))
  - Please note that the cache was previously updated by the subscriptions only. With this change in the PR, we now update the cache when a form/checklist is submitted + when an activity is updated from subscription.
- We also had a tech debt that we are using indices to navigate through the activities instead of their IDs. This PR also includes a small refactor that addresses this change.

### CODE REVIEW

File: src/hooks/activityNavigation/ActivityProvider.tsx [hunk 1]
```diff
@@ -9,33 +9,44 @@ interface ActivityProviderProps {
...
-    if (
-      activities.length === 0 ||
-      activities[currentActivity].status === ActivityStatus.Active
-    ) {
-      return
-    }
-    const currentActivityId = activities[currentActivity].id
+  useEffect(() => {
+    findNextActiveActivity()
+  }, [])
```
Comment (rahulkeerthi) [inline src/hooks/activityNavigation/ActivityProvider.tsx:24]: what does this do? findNextActiveActivity function just returns an array

File: src/hooks/activityNavigation/ActivityProvider.tsx [hunk 2]
```diff
@@ -9,33 +9,44 @@ interface ActivityProviderProps {
...
     )
+  }
 
-    if (nextActivityIndex === -1 || currentActivity === nextActivityIndex) {
+  const handleSetCurrent = () => {
+    const currentActivity = findCurrentActivity()
+
+    // if current activity is still active, then do not move
+    // to the next activity unless it is marked as completed
+    if (currentActivity?.status === ActivityStatus.Active) {
```
Comment (rahulkeerthi) [inline src/hooks/activityNavigation/ActivityProvider.tsx:42]: I can think of two (broad) reasons why currentActivity will be undefined:
1. currentActivityId has not been set yet (is ``) so cannot be found in activities array
2. activities array is empty so currentActivityId cannot be found

in both cases we are doing work even though we don't need to, so we can check if `currentActivityId` is an empty string or activities array is empty, and return early

this logic can live in the `findCurrentActivity` function itself so it's handled silently, and here in `handleSetCurrent` we can explicitly handle if `currentActivity` is undefined.

**Review by rahulkeerthi — COMMENTED** [review]
I see that you've updated the hooks associated with mutations for messages, forms and checklists. Does the cache update hook need to be used for other user activity types/hooks? e.g. cal.com booking activity?

**Comment by mohsinht** [comment]
> Does the cache update hook need to be used for other user activity types/hooks? e.g. cal.com booking activity?

Yes, I was adding cache update function where we call `handleNavigateToNextActivity()`. It was indeed missing in the extension activities. Thanks for pointing that out

File: src/hooks/activityNavigation/ActivityProvider.tsx [hunk 3]
```diff
@@ -19,11 +19,10 @@ export const ActivityProvider: FC<ActivityProviderProps> = ({
 }) => {
   const [currentActivityId, setCurrentActivityId] = useState<string>('')
 
-  useEffect(() => {
-    findNextActiveActivity()
-  }, [])
-
   const findCurrentActivity = (): Activity | undefined => {
+    if (activities.length === 0 || isNil(currentActivityId)) {
```
Comment (rahulkeerthi) [inline src/hooks/activityNavigation/ActivityProvider.tsx:23]: currentActivityId will never be null or undefined, it's set as an empty string by default in the useState constructor above. i think you want to check that it's an empty string

`    if (activities.length === 0 || currentActivityId === '') {`

alternatively, initiate currenctActivityId as undefined

**Pull request merged**